### PR TITLE
resolve: Turn the "non-empty glob must import something" error into a lint

### LINKED ISSUE #62334: Change "A non-empty glob must import something with the glob's visibility" to be a lint?
Consider the following three examples: A and C are accepted and B has a compilation error. However, the message reported in B looks more like a lint than a compilation error.

A consequence of this error is that adding a non-public function to a module (e.g. the `bar` in B) may break code that imports from that module. This causes surprises when refactoring.

Shouldn't "A non-empty glob must import something with the glob's visibility" be a lint?

The original discussion is here: https://users.rust-lang.org/t/a-non-empty-glob-must-import-something-with-the-globs-visibility-uhhh-okay-but-why

### A
```
pub use self::a::*;
mod a {
    fn foo() {}
}
```

### B
```
// ERROR: A non-empty glob must import something with the glob's visibility
pub use self::b::*;
mod b {
    fn foo() {}
    pub(super) fn bar() {}
}
```

### C
```
// okay: there's a pub item
pub use self::c::*;
mod c {
    fn foo() {}
    pub(super) fn bar() {}
    pub fn baz() {}
}
```

### PR DESCRIPTION
**Pull request by traxys**
This fixes #62334 by changing the error to a lint warning the glob. I changed the test but I'm very unsure of what I did as I do not know how to correctly check for the warning

### CODE REVIEW

**Comment by fpoli** [comment]
To update the test I think you only need to replace the `//~ ERROR` with `//~ WARNING`. The annotation syntax is documented in the rustc-guide: https://rust-lang.github.io/rustc-guide/tests/adding.html#error-annotations.

**Comment by petrochenkov** [comment]
If the warning is not reported due to compilation being stopped too early, then the test can be split into two tests - one with errors and another with warnings only.

**Comment by traxys** [comment]
I will need to check this evening, When running rustc I see a warning but the test does pass if I just remove the `//~ ERROR` without looking for a warning.

**Comment by traxys** [comment]
I fixed the test but because the tests in import are run with `-A unused` I needed to add a `#![warn(unused)]`, I don't know if it is the right thing to do

**Comment by petrochenkov** [comment]
>I needed to add a #![warn(unused)], I don't know if it is the right thing to do

Yep, that's the right thing to do.
You could use more specific `warn(unused_imports)` though.

File: src/test/ui/imports/reexports.rs [hunk 1]
```diff
@@ -1,16 +1,22 @@
+#![warn(unused)]
+
+// ignore-tidy-linelength
```
Comment (petrochenkov) [inline src/test/ui/imports/reexports.rs]: This is better avoided, the `//~ ERROR` or `//~ WARNING` annotations can be placed on the line below the error if necessary.
```rust
pub use super::*;
//~^ WARNING must import ...
```

**Comment by petrochenkov** [comment]
r? @petrochenkov 

This probably doesn't need a full FCP, but I'm still going to ping the lang team.
Before that I need to find the rationale for the existing behavior though (why is it an error), I remember it was discussed before.

**Comment by Centril** [comment]
> This probably doesn't need a full FCP, but I'm still going to ping the lang team.

Nominated. :) We'll discuss the aspect of FCP/not & what we think. Having the rationale ready by then would be great.

File: src/test/ui/imports/reexports.rs (hunk 1 above)
Comment (traxys) [reply src/test/ui/imports/reexports.rs]: I have made the changes locally but the I am not able to run the test, I'll be able on sunday and I'll commit them after

**Comment by petrochenkov** [comment]
So, I couldn't find where exactly we discussed this with jseyfried, but the logic behind this is that
- What imports really do is 1) pattern matching and 2) filtering by visibility.
- `use prefix::foo` matches all items named `foo` in the `prefix`, there may be from 0 to 3 of them of them due to multiple namespaces.
- `use prefix::*` matches all items in the `prefix`, there may be from 0 to infinity of them.
- `use prefix::foo`/`use prefix::*` then filter away items that are not visible from the import's location.
- if `use prefix::foo` imports 0 items after the filtering, we report a "private item" or an "unresolverd import" error, which is user friendly, but doesn't precisely reflect what happens. Instead, we could import nothing and let `use prefix::foo` be an unused import, but that's less intuitive and also harmful for the import resolution convergence (we can make progress based on the "single import always introduces a name or produces an error" guarantee).
- applying the same logic to glob imports we get this the "non-empty glob must import something" which is basically an equivalent of the "private item" errors for single imports.

TLDR: The error is reported by analogy with single imports.

However, this analogy doesn't give as much, and it also may be expected from single and glob imports to behave differently by analogy with e.g. regexps - `.*` can match an empty string, but `foo` cannot.
So we can abandon it and make glob imports that import nothing always be valid (but perhaps warned about).

File: src/librustc_resolve/resolve_imports.rs [hunk 2]
```diff
@@ -973,7 +973,7 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                    max_vis.get() != ty::Visibility::Invisible && // Allow empty globs.
                    !max_vis.get().is_at_least(directive.vis.get(), &*self) {
                     let msg = "A non-empty glob must import something with the glob's visibility";
```
Comment (petrochenkov) [inline src/librustc_resolve/resolve_imports.rs]: This should probably be reworded into something without "must".
Like "this glob import doesn't import anything because all the candidates are too private" or something.
(Also, errors and warnings conventionally start with a lowercase letter.)

Comment (traxys) [reply src/librustc_resolve/resolve_imports.rs]: Maybe the other way "no candidate is public enough" ?

**Comment by nikomatsakis** [comment]
So we discussed this in the @rust-lang/lang meeting and felt generally 👍 on warning and not error, except that we were also kind of confused about exactly what scenario we're even talking about. 

This is specifically referring to the case where:

* there is a `pub(V) use` in module X
* there are items that are visible to X, but they do not have visibility `V`
    * in which case, we currently get an error, but we would not give a warning

This all seems fine to us. However, there was one question we wanted clarification on, @petrochenkov. You wrote that the error in the case of an explicit import is helpful for ensuring convergence:

> we can make progress based on the "single import always introduces a name or produces an error" guarantee

Presumably, this doesn't apply to globs?

File: src/librustc_resolve/resolve_imports.rs (hunk 2 above)
Comment (nikomatsakis) [inline src/librustc_resolve/resolve_imports.rs]:
```suggestion
                    let msg = "a non-empty glob must import something with the glob's visibility";
```

we usually make error messages lower-case :)

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
One minor nit (pre-existing, but hey)

File: src/librustc_resolve/resolve_imports.rs (hunk 2 above)
Comment (traxys) [reply src/librustc_resolve/resolve_imports.rs]: In a comment @petrochenkov told we could write something along the lines of "this glob import doesn't import anything because all the candidates are too private", and I asked on "because no candidate is public enough". Should I change just the capital, or change the message to one of the two proposals ?

**Comment by traxys** [comment]
The only thing for wich I have not seen any response is the concern about the warning message, if it should not be changed to "all items are too private" or "no item is public enough"

**Comment by petrochenkov** [comment]
@nikomatsakis 
>we can make progress based on the "single import always introduces a name or produces an error" guarantee
>Presumably, this doesn't apply to globs?

Yes, it doesn't apply to globs.
For single imports we also know *what* name they introduce, for globs the knowledge about some *unknown* name being introduced is pretty useless.

Also, this error was introduced together with `pub(path)`, and glob could introduce 0 names before that, and jseyfried didn't want to break them, so globs can introduce 0 names now as well, e.g.
```rust
mod m {}
use m::*;

fn main() {}
```

File: src/librustc_resolve/resolve_imports.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustc_resolve/resolve_imports.rs]: >Maybe the other way "no candidate is public enough" ?

Sounds good.

Also, "doesn't import anything" -> "doesn't reexport anything", because the does actually import something, but it is not visible to outer modules as stated.

**Comment by petrochenkov** [comment]
Disregard all my comments about progress, I misremembered what this error is about.
It's basically a "private-in-public" errror for globs, this kind of error can be relaxed and turned into a warning for single imports as well.

File: src/librustc_resolve/resolve_imports.rs [hunk 3]
```diff
@@ -972,8 +972,9 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                 if !is_prelude &&
                    max_vis.get() != ty::Visibility::Invisible && // Allow empty globs.
                    !max_vis.get().is_at_least(directive.vis.get(), &*self) {
-                    let msg = "A non-empty glob must import something with the glob's visibility";
-                    self.r.session.span_err(directive.span, msg);
+                    let msg =
+                        "this glob doesn't reexport anything because no canditate is public enough";
```
Comment (petrochenkov) [inline src/librustc_resolve/resolve_imports.rs]:
```suggestion
                        "glob import doesn't reexport anything because no candidate is public enough";
```

File: src/test/ui/imports/reexports.rs (hunk 1 above)
Comment (petrochenkov) [reply src/test/ui/imports/reexports.rs]: The message after `//~^ WARNING` doesn't necessary need to be the full output, any prefix like `//~^ WARNING this glob doesn't reexport anything` will match too.
Using a prefix is preferable to `// ignore-tidy-linelength`, full messages are tested by the `.stderr` files anyway.

**Comment by petrochenkov** [comment]
Also, could you squash the commits into one?

**Comment by traxys** [comment]
Squash all the commits of the pull request together ?

**Comment by nikomatsakis** [comment]
@traxys the usual technique is to use `git rebase -i` and convert all but the first commit to "squash" (and then edit the message)

**Comment by eddyb** [comment]
You can use `fixup` instead of `squash` if you don't want to concatenate the messages, btw (they're otherwise equivalent).
